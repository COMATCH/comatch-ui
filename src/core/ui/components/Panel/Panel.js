import React, { Component } from 'react';
import PropTypes from 'prop-types';
import noop from 'lodash/noop';
import classNames from 'classnames';
import { SlideDown } from 'react-slidedown';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronUp, faChevronDown } from '@fortawesome/free-solid-svg-icons';

import 'react-slidedown/lib/slidedown.css';
import './Panel.scss';

export class Panel extends Component {
    static propTypes = {
        children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
        className: PropTypes.string,
        controls: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
        collapsible: PropTypes.bool,
        expanded: PropTypes.bool,
        disabled: PropTypes.bool,
        id: PropTypes.string,
        onClick: PropTypes.func,
        secondary: PropTypes.bool,
        title: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node, PropTypes.string]),
        titleColor: PropTypes.oneOf(['black', 'gray', 'blue', 'green']),
    };

    static defaultProps = {
        children: null,
        className: null,
        controls: null,
        collapsible: false,
        expanded: false,
        disabled: false,
        id: null,
        onClick: noop,
        secondary: false,
        title: '',
        titleColor: 'black',
    };

    state = {
        expanded: this.props.expanded,
    };

    componentDidUpdate(prevProps) {
        if (this.props.expanded !== prevProps.expanded) {
            this.setState({ expanded: this.props.expanded });
        }
    }

    toggleOpen = () => {
        this.setState((state) => ({ expanded: !state.expanded }));
    };

    handleClick = () => {
        if (this.props.collapsible) {
            this.toggleOpen();
        }
        this.props.onClick();
    };

    renderContents() {
        const { children, collapsible } = this.props;
        const { expanded } = this.state;

        return collapsible ? (
            <SlideDown className="Panel__contents">{expanded ? children : null}</SlideDown>
        ) : (
            <div className="Panel__contents">{children}</div>
        );
    }

    renderHeaderControls() {
        const { controls, collapsible } = this.props;
        const { expanded } = this.state;

        return (
            <div className="Panel__controls">
                {controls}
                {collapsible && (
                    <FontAwesomeIcon className="Panel__chevron" icon={expanded ? faChevronUp : faChevronDown} />
                )}
            </div>
        );
    }

    renderHeader() {
        const { controls, collapsible, title, titleColor } = this.props;
        const titleClasses = classNames('Panel__title', titleColor);

        if (controls !== null || collapsible) {
            return (
                <header className="Panel__header" onClick={this.handleClick}>
                    {this.secondary ? (
                        <h6 className={titleClasses}>{title}</h6>
                    ) : (
                        <h4 className={titleClasses}>{title}</h4>
                    )}
                    {this.renderHeaderControls()}
                </header>
            );
        }
        return (
            title && (
                <header className="Panel__header" onClick={this.handleClick}>
                    {this.secondary ? (
                        <h6 className={titleClasses}>{title}</h6>
                    ) : (
                        <h4 className={titleClasses}>{title}</h4>
                    )}
                </header>
            )
        );
    }

    render() {
        const { children, className, controls, collapsible, disabled, id, secondary } = this.props;
        const { expanded } = this.state;

        const renderContent = disabled === false && children !== null;

        const classes = classNames('Panel', className, {
            collapsible,
            disabled,
            expanded,
            primary: !secondary,
            secondary,
            'without-content': children === null || disabled,
            'with-controls': controls !== null,
        });

        return (
            <section id={id} className={classes}>
                {this.renderHeader()}
                {renderContent && this.renderContents()}
            </section>
        );
    }
}
