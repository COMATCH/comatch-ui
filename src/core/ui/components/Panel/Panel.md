
## Panel

From [`Panel.js`](Panel.js)

The `Panel` is a way to organize and present content in the different pages.

prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**children** | `Union<Array[]<ReactNode> \| ReactNode>` | `null` | :x: | 
**className** | `String` | `null` | :x: | Additional class names
**collapsible** | `Boolean` | `false` | :x: | If true, the `Panel` can expand and collapse its contents. Starts out as collapsed in the first render.
**controls** | `Union<Array[]<ReactNode> \| ReactNode>` | `null` | :x: | Some buttons, links or other HTML content to be placed on the right part of the header next to the title.
**disabled** | `Boolean` | `false` | :x: | 
**expanded** | `Boolean` | `false` | :x: | If true, starts out as expanded in the first render
**id** | `String` | `null` | :x: | HTML id attribute
**onClick** | `Function` | `noop` | :x: | 
**secondary** | `Boolean` | `false` | :x: | Alternative styling
**title** | `Union<Array[]<ReactNode> \| ReactNode \| String>` | `''` | :x: | Heading in the header of the `Panel`
**titleColor** | `Enum('black','gray','blue','green')` | `'black'` | :x: | 



