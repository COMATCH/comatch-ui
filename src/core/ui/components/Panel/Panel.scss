@import '../../../styles/variables.scss';

.Panel {
    background-color: $co-color-white;
    border-radius: 3px;
    border: 1px solid $co-color-mid-gray;
    margin: 10px 0;
    padding: 20px;

    &.with-controls {
        padding: 14.6px 20px;
    }

    // Header in case controls or collapsible are/is set
    &__header {
        display: flex;
        margin-top: -20px;
        margin-left: -20px;
        margin-right: -20px;

        .Panel__title,
        .Panel__controls {
            display: flex;
            align-items: center;
        }

        .Panel__title {
            flex-grow: 1;
        }

        .Panel__controls {
            padding-left: 20px;
            flex-shrink: 1;
            * {
                position: relative;
            }
            * ~ * {
                margin-left: 10px;
            }
        }
    }

    // <h6> heading of Panel
    &__title {
        margin: 0;

        &.green {
            color: $co-color-green;
        }
        &.blue {
            color: $co-color-primary;
        }
        &.gray {
            color: $co-color-mid-gray;
        }
    }

    &__separator {
        background-color: $co-color-mid-gray;
        height: 1px;
    }

    &__separator,
    .Separator {
        margin: 20px -20px;
    }

    // Create whitespace between subsequent Panels
    & ~ & {
        margin-top: 0;
    }

    // Stylistic variations
    &.primary {
        .Panel__header {
            min-height: 50px;
            margin-bottom: 20px;
            padding: 20px 20px 0 20px;
        }
        .Panel__title {
            font-size: $co-font-xl;
            font-weight: normal;
            line-height: $co-line-height-xl;
        }
    }

    &.secondary {
        .Panel__header {
            min-height: 40px;
            margin-bottom: 20px;
            padding: 5px 20px;
            position: relative;

            &:before {
                background-color: $co-color-light-gray;
                border-top-left-radius: 3px;
                border-top-right-radius: 3px;
                border-bottom: 1px solid $co-color-mid-gray;
                content: '';
                display: block;
                width: 100%;
                height: 100%;
                position: absolute;
                top: 0;
                left: 0;
            }
        }
        .Panel__title {
            font-size: $co-font-md;
            line-height: $co-line-height-md;
            font-weight: bold;
            position: relative;
        }
    }

    // States
    &.collapsible {
        .Panel__header {
            cursor: pointer;
            // For headers of all collapsible panels, margin-bottom will animate
            transition: margin-bottom 250ms ease-out;
            will-change: margin-bottom;
            // Setting translateZ forces GPU accelerated animations improving mobile performance
            transform: translateZ(0);
        }

        // For primary styled panels that are collapsible
        &.primary {
            .Panel__header {
                margin-bottom: 0;
            }
        }

        // For secondary styled panels that are collapsible
        &.secondary {
            .Panel__header {
                margin-bottom: -20px;

                &:before {
                    border-bottom-right-radius: 3px;
                    border-bottom-left-radius: 3px;
                    border-bottom: none;
                }
            }
        }
    }

    &.expanded {
        &.primary {
            .Panel__header {
                margin-bottom: 20px;
            }
        }

        &.secondary {
            .Panel__header {
                margin-bottom: 20px;

                &:before {
                    border-bottom-right-radius: 0;
                    border-bottom-left-radius: 0;
                    border-bottom: 1px solid $co-color-mid-gray;
                }
            }
        }
    }

    &.without-content {
        .Panel__header {
            margin-bottom: 0;
        }
    }

    &.disabled {
        .Panel__header {
            color: $co-color-mid-gray;
        }
    }
}

.react-slidedown.Panel__contents {
    transition-duration: 250ms;
    transition-timing-function: ease-out;
}
