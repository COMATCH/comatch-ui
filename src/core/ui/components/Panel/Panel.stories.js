import React from 'react';
import Highlight from 'react-highlight';
import { storiesOf } from '@storybook/react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCheck } from '@fortawesome/free-solid-svg-icons';
import { Panel } from './Panel';
import { Button } from '../Button/Button';

const panelText = `Lorem ipsum dolor sit amet, consectetur adipiscing elit.
Pellentesque porttitor efficitur enim, vitae blandit risus hendrerit non.
Donec in dignissim felis. Aenean eget risus non est suscipit finibus.
Fusce sed ipsum dapibus mi feugiat dapibus vel vel libero. Phasellus varius.`;

const controls = [<Button key={1} text="Ghost" ghost />, <Button key={2} text="Full" />];
const control = <Button text="Ghost" ghost />;

const titleWithIcon = (
    <span>
        <FontAwesomeIcon icon={faCheck} /> Success
    </span>
);

storiesOf('Panel', module)
    .add('primary style with title', () => (
        <>
            <Panel title="A panel title">{panelText}</Panel>
            <Highlight className="html">{`<Panel title="A panel title">{panelText}</Panel>`}</Highlight>
        </>
    ))
    .add('primary style with title and without content', () => (
        <>
            <Panel title="A panel title" />
            <Highlight className="html">{`<Panel title="A panel title" />`}</Highlight>
        </>
    ))
    .add('primary style with green title color', () => (
        <>
            <Panel title="A panel title" titleColor="green">
                {panelText}
            </Panel>
            <Highlight className="html">
                {`<Panel title="A panel title" titleColor="green">{panelText}</Panel>`}
            </Highlight>
        </>
    ))
    .add('primary style with blue title color', () => (
        <>
            <Panel title="A panel title" titleColor="blue">
                {panelText}
            </Panel>
            <Highlight className="html">
                {`<Panel title="A panel title" titleColor="blue">{panelText}</Panel>`}
            </Highlight>
        </>
    ))
    .add('primary style with gray title color', () => (
        <>
            <Panel title="A panel title" titleColor="gray">
                {panelText}
            </Panel>
            <Highlight className="html">
                {`<Panel title="A panel title" titleColor="gray">{panelText}</Panel>`}
            </Highlight>
        </>
    ))
    .add('primary style with title and a control', () => (
        <>
            <Panel title="A panel title" controls={control} primary>
                {panelText}
            </Panel>
            <Highlight className="html">
                {`<Panel title="A panel title" controls={control} primary>{panelText}</Panel>`}
            </Highlight>
        </>
    ))
    .add('primary style with title and controls', () => (
        <>
            <Panel title="A panel title" controls={controls} primary>
                {panelText}
            </Panel>
            <Highlight className="html">
                {`<Panel title="A panel title" controls={controls} primary>{panelText}</Panel>`}
            </Highlight>
        </>
    ))
    .add('collapsible with primary style and blue title', () => (
        <>
            <Panel title="A panel title" titleColor="blue" collapsible>
                {panelText}
            </Panel>
            <Highlight className="html">
                {`<Panel title="A panel title" titleColor="blue" collapsible>{panelText}</Panel`}
            </Highlight>
        </>
    ))
    .add('collapsible with primary style and without content', () => (
        <>
            <Panel title="A panel title" collapsible />
            <Highlight className="html">{`<Panel title="A panel title" collapsible />`}</Highlight>
        </>
    ))
    .add('collapsible with primary style and huge title', () => (
        <>
            <Panel
                title="A very long and exhausting to read panel title that might as well replace the body text"
                collapsible
            >
                {panelText}
            </Panel>
            <Highlight className="html">
                {`<Panel` +
                    `\n\ttitle="A very long and exhausting to read panel` +
                    ` title that might as well replace the body text"` +
                    `\n\tcollapsible` +
                    `\n>` +
                    `\n\t{panelText}` +
                    `\n</Panel>`}
            </Highlight>
        </>
    ))
    .add('collapsible and initially expanded with primary style and blue title', () => (
        <>
            <Panel title="A panel title" titleColor="blue" collapsible expanded>
                {panelText}
            </Panel>
            <Highlight className="html">{`<Panel title="A panel title" collapsible />`}</Highlight>
        </>
    ))
    .add('collapsible with primary style and blue title with icon', () => (
        <>
            <Panel title={titleWithIcon} titleColor="blue" collapsible>
                {panelText}
            </Panel>
            <Highlight className="html">
                {'<Panel title={titleWithIcon} titleColor="blue" collapsible>\n\t{panelText}\n</Panel>'}
            </Highlight>
        </>
    ))
    .add('collapsible with primary style, title and controls', () => (
        <>
            <Panel title="A panel title" controls={controls} primary collapsible>
                {panelText}
            </Panel>
            <Highlight className="html">
                {`<Panel title="A panel title" controls={controls} primary collapsible>` +
                    `\n\t{panelText}` +
                    `\n</Panel>`}
            </Highlight>
        </>
    ))
    .add('secondary style with title', () => (
        <>
            <Panel title="A panel title" secondary>
                {panelText}
            </Panel>
            <Highlight className="html">{'<Panel title="A panel title" secondary>\n\t{panelText}\n</Panel>'}</Highlight>
        </>
    ))
    .add('secondary style with title and without content', () => (
        <>
            <Panel title="A panel title" secondary />
            <Highlight className="html">{`<Panel title="A panel title" secondary />`}</Highlight>
        </>
    ))
    .add('secondary style with title and controls', () => (
        <>
            <Panel title="A panel title" controls={controls} secondary>
                {panelText}
            </Panel>
            <Highlight className="html">
                {'<Panel title="A panel title" controls={controls} secondary>\n\t{panelText}\n</Panel>'}
            </Highlight>
        </>
    ))
    .add('collapsible with secondary style and title', () => (
        <>
            <Panel title="A panel title" collapsible secondary>
                {panelText}
            </Panel>
            <Highlight className="html">
                {'<Panel title="A panel title" collapsible secondary>\n\t{panelText}\n</Panel>'}
            </Highlight>
        </>
    ))
    .add('collapsible with secondary style and huge title', () => (
        <>
            <Panel
                title="A very long and exhausting to read panel title that might as well replace the body text"
                collapsible
                secondary
            >
                {panelText}
            </Panel>
            <Highlight className="html">
                {`<Panel` +
                    `\n\ttitle="A very long and exhausting to read panel title` +
                    ` that might as well replace the body text"` +
                    `\n\tcollapsible` +
                    `\n\tsecondary` +
                    `\n>` +
                    `\n\t{panelText}` +
                    `\n</Panel>`}
            </Highlight>
        </>
    ))
    .add('collapsible initially expanded with secondary style and title', () => (
        <>
            <Panel title="A panel title" collapsible secondary expanded>
                {panelText}
            </Panel>
            <Highlight className="html">
                {'<Panel title="A panel title" collapsible secondary expanded>\n\t{panelText}\n</Panel>'}
            </Highlight>
        </>
    ))
    .add('collapsible with secondary style, title and controls', () => (
        <>
            <Panel title="A panel title" controls={controls} collapsible secondary>
                {panelText}
            </Panel>
            <Highlight className="html">
                {'<Panel title="A panel title" controls={controls} collapsible secondary>' +
                    '\n\t{panelText}' +
                    '\n</Panel>'}
            </Highlight>
        </>
    ))
    .add('without title', () => (
        <>
            <Panel>{panelText}</Panel>
            <Highlight className="html">{`<Panel>{panelText}</Panel>`}</Highlight>
        </>
    ))
    .add('disabled', () => (
        <>
            <Panel title="A panel title" disabled />
            <Highlight className="html">{`<Panel title="A panel title" disabled />`}</Highlight>
        </>
    ));
