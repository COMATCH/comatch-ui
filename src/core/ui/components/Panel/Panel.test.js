/* eslint-env mocha */

import React from 'react';
import { shallow } from 'enzyme';
import { expect } from 'chai';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { SlideDown } from 'react-slidedown';
import { lorem } from 'faker';
import { Panel } from './Panel';
import '../../../config/tests/setup';

describe('Panel', () => {
    it('should render a Panel with default props', () => {
        const item = shallow(<Panel />);
        // should have default classes .Panel.primary.without-content
        expect(item.hasClass('Panel')).to.equal(true);
        expect(item.hasClass('primary')).to.equal(true);
        expect(item.hasClass('without-content')).to.equal(true);
    });
    it('should render a Panel with a title', () => {
        const item = shallow(<Panel title="Title" />);
        // should have a header
        const header = item.find('header');
        expect(header.length).to.equal(1);
        expect(header.find('h4').length).to.equal(1); // should contain primary heading
    });
    it('should render a Panel with controls', () => {
        const item = shallow(<Panel title={lorem.word()} controls={<button>button</button>} />);
        expect(item.hasClass('with-controls')).to.equal(true);
        expect(item.find('.Panel__controls'));
    });
    it('should render a collapsible Panel that is not expanded', () => {
        const item = shallow(
            <Panel title={lorem.word()} collapsible>
                {lorem.word()}
            </Panel>,
        );
        expect(item.hasClass('collapsible')).to.equal(true);
        expect(item.find('.Panel__controls').find(FontAwesomeIcon).length).to.equal(1);

        const slideDown = item.find(SlideDown);
        expect(slideDown.length).to.equal(1);
        expect(slideDown.children().length).to.equal(0);
    });
    it('should render a collapsible Panel that is expanded', () => {
        const item = shallow(
            <Panel title={lorem.word()} collapsible expanded>
                {lorem.word()}
            </Panel>,
        );
        expect(item.hasClass('collapsible')).to.equal(true);
        expect(item.hasClass('expanded')).to.equal(true);

        const slideDown = item.find(SlideDown);
        expect(slideDown.length).to.equal(1);
        expect(slideDown.children().length).to.equal(1);
    });
    it('should render a disabled Panel', () => {
        const item = shallow(
            <Panel title={lorem.word()} disabled>
                {lorem.word()}
            </Panel>,
        );
        expect(item.hasClass('disabled')).to.equal(true);
        expect(item.find('.Panel__contents').length).to.equal(0); // contents shouldn't be rendered
    });
});
