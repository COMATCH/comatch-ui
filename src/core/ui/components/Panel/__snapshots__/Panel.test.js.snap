// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel should render correctly 1`] = `
<section
  className="Panel primary without-content"
  id={null}
>
  
</section>
`;
